morph_tool/morphdb.py

```diff
@@ -0,0 +1,346 @@
+'''Provides two classes to work with neurondb files
+The main use is to provide a neuronDB loader and a method to retrieve
+information as a dataframe.
+
+Example:
+
+old_release = MorphDB.from_neurondb(Path("/realease/one/neuronDB.xml"), label='old-release')
+new_release = MorphDB.from_neurondb(Path("/realease/two/neuronDB.xml"), label='new-release')
+total = old_release + new_release
+
+# neurondb information
+print(total.df)
+
+# combined information (neurondb + morph_stats)
+print(total.features({'neurite': {'section_lengths': ['max']}}))
+'''
+import logging
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Union
+
+import pandas as pd
+import xmltodict
+from neurom.apps.morph_stats import extract_dataframe
+from toolz import isiterable
+
+from morph_tool.utils import iter_morphology_files
+
+L = logging.getLogger(__name__)
+
+
+# Attributes found in the <repair></repair> block of the XML
+BOOLEAN_REPAIR_ATTRS = [
+    'use_axon',
+    'use_dendrites',
+    'axon_repair',
+    'dendrite_repair',
+    'basal_dendrite_repair',
+    'tuft_dendrite_repair',
+    'oblique_dendrite_repair',
+    'unravel',
+    'use_for_stats',
+]
+
+COLUMNS = ['name', 'mtype', 'msubtype', 'mtype_no_subtype',
+           'layer', 'label', 'path'] + BOOLEAN_REPAIR_ATTRS + ['axon_inputs']
+
+
+def find_morph(folder: Path, stem: str) -> Optional[Path]:
+    '''Returns the path to a morphology in morphology_dir matching stem.
+
+    If no morphology is found, returns None
+    '''
+    for ext in {'.asc', '.ASC', '.h5', '.H5', '.swc', '.SWC'}:
+        path = folder / (stem + ext)
+        if path.exists():
```

Comment:
this is slow for large number of files, I had tried to bypass it, if the extension is known in advance

Reply:
Ah, yes. I forgot to put the `extension` argument back. I will do.

Reply:
In fact, since we are here, we could add a method `.check_files_exist()` which runs this `path.exists()` and complain if one is missing. We could then use this method from times to times, to check nothing magically disappeard from gpfs :P